src/Compilers/CSharp/Portable/BoundTree/DecisionTree.cs

```diff
@@ -0,0 +1,916 @@
+                                                if (error != 0) return error;
+                                            }
+                                        }
+                                    }
+                                    return (byType.Default != null) ? CheckSubsumed(pattern, byType.Default, inputCouldBeNull) : 0;
+                                }
+                            case DecisionTree.DecisionKind.Guarded:
+                                {
+                                    var guarded = (DecisionTree.Guarded)decisionTree;
+                                    return
+                                        (guarded.Guard == null || guarded.Guard.ConstantValue == ConstantValue.True) ? ErrorCode.ERR_PatternIsSubsumed :
+                                        guarded.Default == null ? 0 : CheckSubsumed(pattern, guarded.Default, inputCouldBeNull);
+                                }
+                            default:
+                                throw ExceptionUtilities.UnexpectedValue(decisionTree.Kind);
+                        }
+                    }
+                case BoundKind.DeclarationPattern:
+                    {
+                        var declarationPattern = (BoundDeclarationPattern)pattern;
+                        switch (decisionTree.Kind)
+                        {
+                            case DecisionTree.DecisionKind.ByValue:
+                                {
+                                    // A declaration pattern is only subsumed by a value pattern if all of the values are accounted for.
+                                    // For example, when switching on a bool, do we handle both true and false?
+                                    // For now, we do not handle this case. Also, this provides compatibility with previous compilers.
+                                    if (inputCouldBeNull) return 0; // null could never be handled by a value decision
+                                    var byValue = (DecisionTree.ByValue)decisionTree;
+                                    if (byValue.Default != null)
+                                    {
+                                        return CheckSubsumed(pattern, byValue.Default, false);
+                                    }
+                                    return 0;
+                                }
+                            case DecisionTree.DecisionKind.ByType:
+                                {
+                                    var byType = (DecisionTree.ByType)decisionTree;
+                                    if (declarationPattern.IsVar && inputCouldBeNull && (byType.WhenNull == null || CheckSubsumed(pattern, byType.WhenNull, inputCouldBeNull) == 0) && (byType.Default == null || CheckSubsumed(pattern, byType.Default, inputCouldBeNull) == 0))
+                                        return 0; // new pattern catches null if not caught by existing WhenNull or Default
```

Comment:
It looks like we are going to get here only when declarationPattern.IsVar. What about the case when it is not Var?